handler/src/main/java/io/netty5/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -517,57 +528,89 @@ private int writePlaintextData(final ByteBuffer src, int len) {
+                SslWrite write = new SslWrite(ssl, len);
+                buf.forEachReadable(0, write);
+                sslWrote = write.sslWrote;
                 if (sslWrote > 0) {
                     src.position(pos + sslWrote);
                 } else {
                     src.position(pos);
                 }
-            } finally {
-                buf.release();
             }
         }
         return sslWrote;
     }
 
+    static final class SslWrite implements ReadableComponentProcessor<RuntimeException> {
+        private final long ssl;
+        private final int len;
+        int sslWrote;
+
+        SslWrite(long ssl, int len) {
+            this.ssl = ssl;
+            this.len = len;
+        }
+
+        @Override
+        public boolean process(int index, ReadableComponent component) {
+            sslWrote = SSL.writeToSSL(ssl, component.readableNativeAddress(), len);
+            return false;
+        }
+    }
+
     /**
      * Write encrypted data to the OpenSSL network BIO.
      */
-    private ByteBuf writeEncryptedData(final ByteBuffer src, int len) throws SSLException {
+    private Buffer writeEncryptedData(final ByteBuffer src, int len) throws SSLException {
+        assert len > 0 && src.remaining() > 0;
         final int pos = src.position();
         if (src.isDirect()) {
             SSL.bioSetByteBuffer(networkBIO, bufferAddress(src) + pos, len, false);
         } else {
-            final ByteBuf buf = alloc.directBuffer(len);
+            Buffer buf = alloc.allocate(len);
             try {
-                final int limit = src.limit();
-                src.limit(pos + len);
-                buf.writeBytes(src);
-                // Restore the original position and limit because we don't want to consume from `src`.
-                src.position(pos);
-                src.limit(limit);
-
-                SSL.bioSetByteBuffer(networkBIO, memoryAddress(buf), len, false);
+                buf.writeBytes(src.array(), src.arrayOffset() + pos, len);
+                SslBioSetByteBuffer setBB = new SslBioSetByteBuffer(networkBIO, len);
+                buf.forEachReadable(0, setBB);
                 return buf;
             } catch (Throwable cause) {
-                buf.release();
+                buf.close();
                 throw cause;
             }
         }
         return null;
     }
 
+    static final class SslBioSetByteBuffer implements ReadableComponentProcessor<RuntimeException>,
+                                                      WritableComponentProcessor<RuntimeException> {
+        private final long networkBIO;
+        private final int len;
+
+        SslBioSetByteBuffer(long networkBIO, int len) {
+            this.networkBIO = networkBIO;
+            this.len = len;
+        }
+
+        @Override
+        public boolean process(int index, ReadableComponent component) {
+            SSL.bioSetByteBuffer(networkBIO, component.readableNativeAddress(), len, false);
```

Comment:
I wonder if we need to somehow ensure that the `ReadableComponent` and so its `Buffer` is not GC'ed until we not use the native pointer anymore. Otherwise wee could end up with a crash. Same comment is true for below.

Reply:
In this case it's up to the caller to ensure. The buffer instance is returned to them.